src/hasSubobject.ts

```diff
@@ -0,0 +1,57 @@
+import { equals } from "./equals";
+import { purry } from "./purry";
+
+/**
+ * Checks if `subObject` is a subobject of `object`, which means for every
+ * property and value in `subObject`, there's the same property in `object`
+ * with an equal value. Equality is checked with `equals`.
+ * @param object the object to test
+ * @param subObject the subobject to test against
+ * @signature
+ *    R.hasSubobject(object, subObject)
+ * @example
+ *    R.hasSubobject({ a: 1, b: 2, c: 3 }, { a: 1, c: 3 }) //=> true
+ *    R.hasSubobject({ a: 1, b: 2, c: 3 }, { b: 4 }) //=> false
+ *    R.hasSubobject({ a: 1, b: 2, c: 3 }, {}) //=> true
+ * @dataFirst
+ * @category Object
+ */
+export function hasSubobject<
+  S extends Record<PropertyKey, unknown>,
+  T extends S,
+>(object: T, subObject: S): boolean;
```

Comment:
1. Just thinking out loud here looking at #565 I just sent: this function is a partial deep equality check, e.g. it's the same exact logic as isDeepEquality, with a single relaxation of the `keys.length !== Object.keys(other).length` check; we can enforce that the "subObject" overlaps with the type of the data, and have the function narrow the result instead of returning `boolean`, this will provide more type-safety when a prop name changes between the checked object and the data, and would provide a better type downstream.

2. More generally, i prefer using `object` as the type instead of `Record<PropertyKey, unknown>` as it's more broad and accepts weirder stuff too, and not just really plain simple records.

```suggestion
export function hasSubobject<
T /* extends object? --- we might need it to make the typing work for the impl? */,
S extends T = T,
>(object: T, subObject: S): object is T & S;
```

Reply:
i assume you mean `S extends object` and `T extends S = S`

i'm not sure i see the point to have the type guard... `object` has to have type `T extends S` anyway, so it must be `T & S` right?

Reply:
Maybe I don't understand the type correctly, this is what I'm thinking:
```ts
type A = {
  a?: number;
  b?: string;
};

declare a: A = {};
if (hasSubObject(a, { b: "hello" })) {
  console.log(a);
  //          ^?: { a?: number, b: string }
}

if (hasSubObject(a, { b: "hello" } as const)) {
  console.log(a);
  //          ^?: { a?: number, b: "hello" }
}

// @ts-expect-error
hasSubObject(a, { c: true });
```